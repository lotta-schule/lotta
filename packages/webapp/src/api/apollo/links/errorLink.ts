import { onError } from '@apollo/link-error';

export const createErrorLink = () =>
  onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({ message, locations, path }) => {
        console.error(
          `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
        );
      });
    }
    if (networkError) {
      console.error(`[Network error]: ${networkError}`);
    }
  });
