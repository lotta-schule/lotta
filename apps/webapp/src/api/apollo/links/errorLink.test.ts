import { describe, it, expect, vi, beforeEach } from 'vitest';
import { createErrorLink } from './errorLink';

describe('createErrorLink', () => {
  const consoleErrorSpy = vi.spyOn(console, 'error');

  beforeEach(() => {
    consoleErrorSpy.mockImplementation(() => {});
  });

  afterEach(() => {
    consoleErrorSpy.mockRestore();
  });

  it('should log GraphQL errors to the console', () => {
    const errorLink = createErrorLink();
    const graphQLErrors = [
      {
        message: 'Test GraphQL error',
        locations: [{ line: 1, column: 2 }],
        path: ['testPath'],
      },
    ];

    expect(() =>
      (errorLink as any).onError({ graphQLErrors, networkError: null })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should log network errors to the console', () => {
    const errorLink = createErrorLink();
    const networkError = new Error('Test Network error');

    expect(() =>
      (errorLink as any).onError({ graphQLErrors: null, networkError })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should handle both GraphQL and network errors', () => {
    const errorLink = createErrorLink();
    const graphQLErrors = [
      {
        message: 'Test GraphQL error',
        locations: [{ line: 1, column: 2 }],
        path: ['testPath'],
      },
    ];
    const networkError = new Error('Test Network error');

    expect(() =>
      (errorLink as any).onError({ graphQLErrors, networkError })
    ).toThrowErrorMatchingSnapshot();
  });
});
