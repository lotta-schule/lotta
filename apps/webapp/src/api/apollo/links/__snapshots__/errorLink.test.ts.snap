// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createErrorLink > should handle both GraphQL and network errors 1`] = `
{
  "graphQLErrors": [
    {
      "locations": [
        {
          "column": 2,
          "line": 1,
        },
      ],
      "message": "Test GraphQL error",
      "path": [
        "testPath",
      ],
    },
  ],
  "networkError": [Error: Test Network error],
}
`;

exports[`createErrorLink > should log GraphQL errors to the console 1`] = `
{
  "graphQLErrors": [
    {
      "locations": [
        {
          "column": 2,
          "line": 1,
        },
      ],
      "message": "Test GraphQL error",
      "path": [
        "testPath",
      ],
    },
  ],
  "networkError": null,
}
`;

exports[`createErrorLink > should log network errors to the console 1`] = `
{
  "graphQLErrors": null,
  "networkError": [Error: Test Network error],
}
`;
